dattri/algorithm/influence_function.py

```diff
@@ -5,99 +5,154 @@
-                name of a parameter and the value is the parameter tensor.
-                TODO: This should be changed to support a list of parameters or
-                    paths for ensembling and memory efficiency.
-            ihvp_solver (str): The solver for inverse hessian vector product
-                calculation, currently we only support "explicit", "cg" and "arnoldi".
-            ihvp_kwargs (Optional[Dict[str, Any]]): Keyword arguments for ihvp solver.
-                calculation, currently we only support "explicit", "cg", "arnoldi",
-                and "lissa".
-            device (str): The device to run the attributor. Default is cpu.
+            index (int): The index of the model parameters. This index
+                is used for ensembling of different trained model
+                parameters.
+            data (Tuple[Tensor]): The test data. Normally this is a tuple
+                of input data and target data, the number of items in the
+                tuple should be aligned in the target function. The tensors'
+                shape follows (1, batchsize, ...).
+
+        Returns:
+            torch.Tensor: The query based on the test data. Normally it is
+                a 2-d dimensional tensor with the shape of
+                (batchsize, num_parameters).
+        """
+        model_params, _ = self.task.get_param(index)
+        return self.task.get_grad_target_func()(model_params, data)
+
+    def generate_train_query(
+        self,
+        index: int,
+        data: Tuple[torch.Tensor, ...],
+    ) -> torch.Tensor:
+        """Calculating the query based on the train data.
+
+        Args:
+            index (int): The index of the model parameters. This index
+                is used for ensembling of different trained model
+                parameters.
+            data (Tuple[Tensor]): The train data. Normally this is a tuple
+                of input data and target data, the number of items in the
+                tuple should be aligned in the target function. The tensors'
+                shape follows (1, batchsize, ...).
+
+        Returns:
+            torch.Tensor: The query based on the train data. Normally it is
+                a 2-d dimensional tensor with the shape of
+                (batchsize, num_parameters).
+        """
+        model_params, _ = self.task.get_param(index)
+        return self.task.get_grad_target_func()(model_params, data)
+
+    @abstractmethod
+    def transformation_on_query(
+        self,
+        index: int,
+        data: Tuple[torch.Tensor, ...],
+        query: torch.Tensor,
+    ) -> torch.Tensor:
+        """Calculate the transformation on the query.
+
+        Inner product attributor calculates the inner product between the
+        test query and the transformation of the test query. This function
+        calculates the transformation of the test query based on the train data.
```

Comment:
transformation of the "train" query based on the train data?
but do we care about what kind of data we used to call this function?